File: BI-TJV/semestral/rest-server/src/test/java/wrzecond/service/TjvAllergenServiceTest.java
package wrzecond.service;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.util.ReflectionTestUtils;
import wrzecond.allergen.TjvAllergenCreateDTO;
import wrzecond.allergen.TjvAllergenReadDTO;
import wrzecond.entity.TjvAllergen;

import java.util.Random;

@SpringBootTest
public class TjvAllergenServiceTest {

    @Autowired
    private TjvAllergenService service;

    @Test
    void toDTO () {
        TjvAllergen allergen = new TjvAllergen("Wheat");
        ReflectionTestUtils.setField(allergen, "id", (new Random()).nextInt());
        TjvAllergenReadDTO dto = new TjvAllergenReadDTO(allergen.getId(), "Wheat");
        Assertions.assertEquals(dto, service.toDTO(allergen));
    }

    @Test
    void fromDTO () {
        TjvAllergenCreateDTO dto = new TjvAllergenCreateDTO("Wheat");
        TjvAllergen allergen = new TjvAllergen("Wheat");
        Assertions.assertEquals(allergen, service.fromDTO(dto));
    }

    @Test
    void merge () {
        TjvAllergen allergen = new TjvAllergen("Wheat"), allergen2 = new TjvAllergen("Milk");
        TjvAllergenCreateDTO dto = new TjvAllergenCreateDTO(null),
                dto2 = new TjvAllergenCreateDTO("Milk");

        // Test
        Assertions.assertEquals(allergen, service.merge(allergen, dto));
        Assertions.assertEquals(allergen2, service.merge(allergen, dto2));
    }

}

File: BI-KOT/semestral/rest-server/src/test/kotlin/wrzecond/service/TjvAllergenServiceTest.kt
package wrzecond.service

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.mockk.mockk
import wrzecond.TjvAllergenCreateDTO
import wrzecond.TjvAllergenReadDTO
import wrzecond.TjvAllergenUpdateDTO
import wrzecond.entity.TjvAllergen
import wrzecond.repository.TjvAllergenRepository

class TjvAllergenServiceTest : StringSpec ({
    val mockRepository = mockk<TjvAllergenRepository>()
    val service = TjvAllergenService(mockRepository)

    "toDTO" {
        val allergen = TjvAllergen("Wheat", 42)
        service.run { allergen.toDTO() } shouldBe TjvAllergenReadDTO(allergen.id, allergen.name)
    }

    "fromDTO" {
        val dto = TjvAllergenCreateDTO("Wheat")
        service.run { dto.toEntity() } shouldBe TjvAllergen(dto.name, 0)
    }

    "merge" {
        val allergen1 = TjvAllergen("Wheat", 1)
        val dto1 = TjvAllergenUpdateDTO(null)
        service.run { allergen1.merge(dto1) } shouldBe allergen1

        val dto2 = TjvAllergenUpdateDTO("Milk")
        service.run { allergen1.merge(dto2) } shouldBe TjvAllergen("Milk", 1)
    }
})